wpilibj/src/main/java/edu/wpi/first/wpilibj/glass/Canvas2d.java

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.wpilibj.glass;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import edu.wpi.first.networktables.FloatArrayPublisher;
+import edu.wpi.first.networktables.NTSendable;
+import edu.wpi.first.networktables.NTSendableBuilder;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.StructArrayPublisher;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+
+public class Canvas2d implements NTSendable, AutoCloseable {
+  private final float[] m_dims = new float[2];
+
+  // TODO: Seperate list of each primitive type, each one with its own NT topic
+  private final List<CanvasLine2d> m_lines = new ArrayList<>();
+
+  private boolean m_sendableInitialized = false;
+  private FloatArrayPublisher m_dimsPub;
+  private StructArrayPublisher<CanvasLine2d> m_linesPub;
+
+  private int m_currentZOrder = 0;
+
+  public Canvas2d(float width, float length) {
+    m_dims[0] = width;
+    m_dims[1] = length;
+  }
+
+  public void draw(CanvasComplexObject2d object) {
+    object.drawOn(this);
+  }
+
+  public void clear() {
+    m_lines.clear();
+  }
+
+  public void drawLine(float x1, float y1, float x2, float y2, float thickness, Color8Bit color) {
+    m_lines.add(new CanvasLine2d(x1, y1, x2, y2, thickness, color, m_currentZOrder++));
+  }
+
+  public void finishFrame() {
+    if (!m_sendableInitialized) {
+      return;
+    }
+    m_linesPub.set(m_lines.toArray(new CanvasLine2d[0]));
```

Comment:
m_lines is never cleared, so doesn't this just keep accumulating lines?

Alternatively, there also be a startFrame() call which clears everything (and resets z-order), and to enforce it gets called, we could add an "m_inFrame" boolean (set in startFrame, cleared in finishFrame) that we check at each draw call to verify we're actually in a frame?

Reply:
My thought was that clear() would serve the place for `startFrame` so that a user could keep drawing on the same canvas if their use case required that. Perhaps renaming `endFrame` to something like `pushFrame`/`publishFrame` would make that more clear, or maybe just require that users restart drawing after every push/finish. What do you think?

Reply:
I'm concerned about not clearing automatically by default, because a naive user will likely not realize what's happening and be surprised when they either run out of memory or consume excessive bandwidth.  If sending an update without clearing is really a desired option, maybe expose it as an optional/defaulted parameter?